pkg/operator/ozone/controller.go

```diff
@@ -0,0 +1,280 @@
+	templateDir := os.Getenv("ROOK_OZONE_TEMPLATE_DIR")
+	if templateDir == "" {
+		templateDir = "/var/lib/ozone-templates"
+	}
+
+	return &Controller{
+		context:     context,
+		templateDir: templateDir,
+	}
+}
+
+// StartWatch watches for instances of ObjectStore custom resources and acts on them
+func (c *Controller) StartWatch(namespace string, stopCh chan struct{}) error {
+	resourceHandlerFuncs := cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.onAdd,
+		UpdateFunc: c.onUpdate,
+		DeleteFunc: c.onDelete,
+	}
+
+	logger.Infof("start watching object store resources in namespace %s", namespace)
+	watcher := opkit.NewWatcher(ObjectStoreResource, namespace, resourceHandlerFuncs, c.context.RookClientset.OzoneV1alpha1().RESTClient())
+	go watcher.Watch(&ozonev1alpha1.ObjectStore{}, stopCh)
+
+	return nil
+}
+
+//The main resource file generation logic from the CRD
+func (c *Controller) generateResources(objectStore *ozonev1alpha1.ObjectStore) error {
+	context, err := flekszible.Initialize(c.templateDir)
+	if err != nil {
+		return err
+	}
+
+	processors := make([]processor.Processor, 0)
+
+	processors = append(processors, &processor.Namespace{
+		Namespace:          objectStore.Namespace,
+		Force:              true,
+		ClusterRoleSupport: true,
+	})
+
+	processors = append(processors, &processor.Image{
+		Image: objectStore.Spec.OzoneVersion.Image,
+	})
+
+	emptyDir, err := processor.CreateTransformation("ozone/emptydir")
+	if err != nil {
+		return err
+	}
+
+	processors = append(processors, emptyDir)
+
+	processors = append(processors, &processor.Replace{
+		Path: data.NewPath("spec", "replicas"),
+		Trigger: processor.Trigger{
+			Definition: data.NodeFromPathValue(data.NewPath("metadata", "name"), "datanode"),
+		},
+		Value: objectStore.Spec.Storage.NodeCount,
+	})
+
+	resources, err := flekszible.Generate(c.templateDir, processors)
+	if err != nil {
+		return err
+	}
+	println(context)
+	for _, resource := range resources {
+		switch kind := resource.Kind(); kind {
+		case "StatefulSet":
+			err = c.applyStatefulSet(resource, objectStore)
+			if err != nil {
+				logger.Errorf("Can't apply StatefulSet %s %s ", resource.Name(), err.Error())
+			}
+		case "Service":
+			err = c.applyService(resource, objectStore)
+			if err != nil {
+				logger.Errorf("Can't apply Service %s %s ", resource.Name(), err.Error())
+			}
+		case "ConfigMap":
+			err = c.applyConfigMap(resource, objectStore)
+			if err != nil {
+				logger.Errorf("Can't apply ConfigMap %s %s ", resource.Name(), err.Error())
+			}
+		default:
+			logger.Errorf("Unhandled resource type: %s", resource.Kind())
+		}
+	}
+	return nil
+}
+
+//helper function to set the owner for k8s resoures
+func (c *Controller) setOwnerRef(objectStore *ozonev1alpha1.ObjectStore, destination *metav1.ObjectMeta) {
+	ownerRef := metav1.OwnerReference{
+		APIVersion: ObjectStoreResource.Version,
+		Kind:       ObjectStoreResource.Kind,
+		Name:       objectStore.Namespace,
```

Comment:
You'll need to set the Name to the name of the object store, not the namespace. This was a critical bug for rook a few months ago where the garbage collector removed rook resources unexpectedly.

Reply:
Ok, I rebased and I found your fix in minio.